### ISSUE TITLE
Running a task within a task

### ISSUE THREAD

**Issue by tomzx** [deployphp/docs#8]
Is there anyway to run a task within another task? I haven't found anything about that in the documentation.

**Comment by elfet** [comment]
Yes, it is not possible. Due to architecture of Deployer. 

If you will be calling tasks from another task, it will be impossible to parallel execution process.

**Comment by tomzx** [comment]
>  it will be impossible to parallel execution process

Why not? You should be able to parallelize at the task level as long as everything is well separated. For instance, I could have a task that runs `composer` while another one runs `grunt` and another one runs `bower`. In my composer task, I should be able to call some other task to `install` composer, `update` it and then run `install`.`

My understanding of the situation is that this is almost similar to having `before` and `after` on tasks. Except in this particular case, I want to run the task in the middle of another one.

If it is not possible, then I think we should work toward making that a possibility.

**Comment by elfet** [comment]
Every task has to be a little as possible to be simple to maintain. You can have a few tasks for composer, grunt, bower. And group them into:
```php
task('build', ['composer', 'grunt', 'bower']);
```

> I want to run the task in the middle of another one.

What if some of sub task fails? 

Current Deployer run tasks like this:
```
task1
↳ server1
↳ server2
task2
↳ server1
↳ server2
...
```
And it's not impossible to put sub tasks on first level.

**Comment by tomzx** [comment]
I thought you used the following parallelization model, which would allow for subtasking.

```
server1
↳ task1
	↳ subtask1
↳ task2
	↳ subtask2
server2
↳ task1
	↳ subtask1
↳ task2
	↳ subtask2
```

**Comment by elfet** [comment]
I was like you described in Deployer 1.0.

New approach is better for parallel execution and server consistency.  Is task fails on one of server, Deployer ends next tasks execution.

**Comment by tomzx** [comment]
Maybe this discussion should be moved to the main project.

I can see the interest in the current way it works, but it also brings a problem. What if a single server has some issue during a step? Then the deployment on all other machines will fail. Obviously, one can remove the broken machine from the pool and deploy again (and have the deployment hopefully succeed).

With the other model, all servers that can complete deployment successfully will. Only broken machines will end up in a state requiring manual intervention.

I guess both case can be interesting. I do not know how difficult it would be to support both, which would give us the most flexibility obviously.

**Comment by elfet** [comment]
> I can see the interest in the current way it works, but it also brings a problem.

Consistency. Deployer keeps consistency between servers, even in parallel mode! If one task fails, subsequent tasks will not be executed on all servers. 

The right solution is combine execution strategy.  It will be possible to implement after v3 release.

**Comment by mrfroasty** [comment]
This works for me:

`task('main:task', function () {
    $mainTasks = [
        'sub_task_1',
        'sub_task_2',
        'sub_task_3'
    ];

    foreach ($mainTasks as $task) {
        invoke($task);
    }
})->desc('Task description');`
